### PR TITLE
code for visual question answering singnet service

### PR DESCRIPTION
**Pull request by noskill**
initial implementation, it's working on local network

### CODE REVIEW

File: services/vqa-service/service.proto [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  
+service VqaService {
+    rpc answer(VqaRequest) returns (VqaResponse);
+}
+
+message VqaRequest {
+    string question = 1;
+    string image_format = 2;
+    bool use_pm = 3;
+    bytes data = 4;
```
Comment (vsbogd) [inline services/vqa-service/service.proto]: may be it is better to name it `image` or `image_data`?

File: services/vqa-service/service.proto [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+message VqaResponse {
+    string message = 1;
```
Comment (vsbogd) [inline services/vqa-service/service.proto]: It is not clear what does `message` mean here, especially taking into account `error_message` is here as well. Should we name it `answer`?

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py [hunk 3]
```diff
@@ -344,11 +353,11 @@ def answerQuestionByImage(self, image, question, use_pm=True) -> QueryProcessing
                 queryInScheme = self.questionConverter.convertToOpencogSchemeURE(relexFormula)
             if queryInScheme is None:
                 self.logger.error('Question was not parsed')
-                return
+                return FailedProcessingData('Question was not parsed')
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: May  be "Unsupported type of the question" here and below ? "Question was not parsed" may mean some internal error, but here only reason is probably unsupported question type. I am not sure, what do you think?

File: services/vqa-service/docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,120 @@
+FROM opencog/vqa
```
Comment (vsbogd) [inline services/vqa-service/docker/Dockerfile:1]: I don't think it makes sense to copy most of workshop docker infrastructure here. Did you want to make a copy of environment for testing purposes?

Comment (noskill) [reply services/vqa-service/docker/Dockerfile:1]: Yes. I don't know yet where vqa service will be deployed.

File: services/vqa-service/service_metadata.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+    "version": 1,
+    "display_name": "ProcessService",
+    "encoding": "proto",
+    "service_type": "grpc",
+    "payment_expiration_threshold": 40320,
+    "model_ipfs_hash": "QmTAXBccCXmN2G6fjzBpME8aaBiLdsp8fNTLAD2mbLCjak",
```
Comment (vsbogd) [inline services/vqa-service/service_metadata.json]: some of the fields like IPFS hashes and even contract addresses here are subjects to change. If you would like to save the infrastructure to deploy service easily it is better to save it as shell script which executes `snet-cli` to deploy the application and has parameters like endpoints and price.

File: services/vqa-service/service_pb2.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
+# Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (vsbogd) [inline services/vqa-service/service_pb2.py]: Autogenerated file should not be kept in repo, I would create `setup.py` which builds Python package using protobuf compiler instead.

File: services/vqa-service/service_pb2_grpc.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
+# Generated by the gRPC Python protocol compiler plugin. DO NOT EDIT!
```
Comment (vsbogd) [inline services/vqa-service/service_pb2_grpc.py]: See https://github.com/singnet/semantic-vision/pull/26/files#r247529142

File: services/vqa-service/snetd.config.json [hunk 8]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (vsbogd) [inline services/vqa-service/snetd.config.json]: See https://github.com/singnet/semantic-vision/pull/26/files#r247528671

File: services/vqa-service/docker/Dockerfile (hunk 4 above)
Comment (vsbogd) [reply services/vqa-service/docker/Dockerfile:1]: I think the better alternative will be a docker image which contains only VQA service itself and you can start it as standalone docker container which exports only gRPC port to interact with VQA service. Then you can use locally installed tools or another image (for instance workshop one) to publish your service to the Kovan or local network and test it. But you don't need to copy local network infrastructure to the VQA repo.

File: services/vqa-service/service.proto [hunk 9]
```diff
@@ -0,0 +1,19 @@
+syntax = "proto3";
+  
+service VqaService {
+    rpc answer(VqaRequest) returns (VqaResponse);
+}
+
+message VqaRequest {
+    string question = 1;
+    string image_format = 2;
```
Comment (vsbogd) [inline services/vqa-service/service.proto]: `image_format` is not used at the server side, may be just remove it?

File: services/vqa-service/docker/Dockerfile (hunk 4 above)
Comment (noskill) [reply services/vqa-service/docker/Dockerfile:1]: That is good idea!

File: services/vqa-service/docker/run.sh [hunk 10]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
```
Comment (vsbogd) [inline services/vqa-service/docker/run.sh]: Seems should be removed as well as part of old docker image.

File: services/vqa-service/docker/start.sh [hunk 11]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
```
Comment (vsbogd) [inline services/vqa-service/docker/start.sh]: Seems should be removed as well as part of old docker image.

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py (hunk 3 above)
Comment (noskill) [reply experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: yes, is formula is correct then queryInScheme would be not None

Comment (vsbogd) [reply experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: Then "Unsupported type of question" would be more clear for the end user I believe.

**Pull request merged**